gluetool/action.py

```diff
@@ -0,0 +1,316 @@
+    of work.
+
+    :param str label: a human-readable string which concisely represents the work done by the ``Action``. The name
+        should be the most general string that describes an interesting class of ``Action`` instances.
+        I.e. ``fetch-url` is better than ``fetch-url-https://foo.com```.
+    :param Action parent: parent ``Action`` - one action can spawn multiple additional "children" actions to
+        achieve its goal, either explicitly or by using instrumented library code.
+    :param dict tags: additional key/value tags of this action, e.g. ``url=https://foo.com``.
+    :param ContextAdapter logger: logger to use for logging purposes.
+    """
+
+    # For each thread, we keep a LIFO of unfinished actions. The topmost one is considered to be
+    # "current".
+    #
+    # Works as long as user keeps only a single "active" action in a thread - if one creates two actions,
+    # side by side, the last one becomes the "current":
+    #
+    # A1, A2 = Action(), Action()
+    #     ^ current action
+    _thread_actions = threading.local()
+
+    @staticmethod
+    def _action_stack():
+        # type: () -> List[Action]
+        """
+        Return current - or create an empty new one - list of unfinished actions of the current thread.
+        """
+
+        if not hasattr(Action._thread_actions, 'stack'):
+            Action._thread_actions.stack = []
+
+        return cast(
+            List[Action],
+            Action._thread_actions.stack
+        )
+
+    @staticmethod
+    def _add_action(action):
+        # type: (Action) -> None
+        """
+        Add action on top of the list of unfinished actions of the current thread.
+        """
+
+        Action._action_stack().append(action)
+
+    @staticmethod
+    def _drop_action(action):
+        # type: (Action) -> None
+        """
+        Drop action from the list of unfinished actions of the current thread.
+        """
+
+        Action._action_stack().remove(action)
+
+    @staticmethod
+    def current_action():
+        # type: () -> Action
+        """
+        Return the top-most - "current" - unfinished action of the current thread.
+        """
+
+        stack = Action._action_stack()
+
+        if not stack:
+            raise RuntimeError('Action stack is empty')
+
+        return stack[-1]
+
+    @staticmethod
+    def set_thread_root(action):
+        # type: (Action) -> None
+        """
+        Initialize list of unfinished action of the current thread with a given transaction.
+
+        When thread starts, its list is obviously empty, therefore :py:meth:`current_action`
+        cannot return anything reasonable. But there probably was an action, e.g. the one in
+        the main thread, which could serve as "current action" for this thread. This method
+        inserts it into the threads list, as the first action.
+        """
+
+        Action._action_stack()[:] = [action]
```

Comment:
a comment for newbies would be nice for this list replacement ...

Reply:
Added.